docs/architecture/adr-060-abci++-phase-i.md

```diff
@@ -0,0 +1,254 @@
+	// mempool, up to maxBytes or until the mempool is empty. The application can
+	// decide to return transactions from its own mempool or from the incoming
+	// TxRecords or some combination of both. The notion of 'available' or 'next'
+	// is defined by the application's mempool implementation.
+	ReapMaxBytes(ctx sdk.Context, txRecords abci.TxRecords, maxBytes int) ([]PrepareTxRecord[T], error)
+	// NumTxs returns the number of transactions currently in the mempool.
+	NumTxs() int
+	// Remove attempts to remove a transaction from the mempool, returning an error
+	// upon failure.
+	Remove(sdk.Context, T) error
+}
+```
+
+We will define an implementation of `Mempool[T MempoolTx]` that will cover a
+majority of application use cases. Namely, it will prioritize transactions by
+priority and transaction sender, allowing for multiple prioritized transactions
+from the same sender. The app-side mempool will be defined as a wrapper around a
+simple priority queue using a max binary heap, along with additional indexes/metadata
+to store senders and their nonces, allowing for simple multi-dimensional
+prioritization (2-ary).
+
+Transaction reaping will essentially happen via a two-phase approach:
+
+1. Reap one or more transactions from the priority queue and collect them into 
+   one of two buffers -- _valid_ and _invalid_.
+2. For transactions that DO NOT violate the nonce validation, they are included
+   in the _valid_ buffer.
+3. For transactions that DO violate the nonce validation, they are included in
+   the _invalid_ buffer.
+4. Continue this process until the desired number of valid transactions are
+   reaped or until the mempool is empty.
+5. Provide Tendermint the list of all transactions from the _valid_ buffer.
+6. Re-insert all transactions, from both buffers, back into app-side mempool.
+   This is to ensure we do not discard transactions from the app-side mempool in
+	 case `ProcessProposal` fails or in case that the proposal, while passing
+	 `ProcessProposal` is not the one decided for that height, i.e. the height took
+	 more than one round.
+
+```go
+type PriorityMempool[T MempoolTx] struct {
+	queue *PriorityQueue[MempoolTx]
+
+	// senders will contain a mapping from tx sender account addresses to all
+	// sequence numbers (nonces) or txs that they have in the app-side mempool.
+	senders map[string][]int64
+
+	// ...
+}
+```
+
+> The `PriorityMempool[T MempoolTx]` implementation will support Options such as
+> limiting the mempool size by a fixed number of bytes.
+
+Previous discussions<sup>1</sup> have come to the agreement that Tendermint will
+perform a request to the application, via `RequestPrepareProposal`, with a certain
+amount of transactions reaped from Tendermint's local mempool. The exact amount
+of transactions reaped will be determined by a local operator configuration.
+This is referred to as the "one-shot approach" seen in discussions.
```

Comment:
If the mempool has more transactions than requested, what determines which ones are reaped?

Reply:
So it's a Tendermint config option?